File: Implementierung/Backend/src/main/java/edu/kit/informatik/dto/mapper/courses/SeatArrangementMapper.java
package edu.kit.informatik.dto.mapper.courses;

import edu.kit.informatik.dto.mapper.IModelDtoMapper;
import edu.kit.informatik.dto.userdata.courses.SeatArrangementDto;
import edu.kit.informatik.model.User;
import edu.kit.informatik.model.userdata.courses.Course;
import edu.kit.informatik.model.userdata.courses.SeatArrangement;
import edu.kit.informatik.model.userdata.interactions.Participant;
import edu.kit.informatik.model.userdata.rooms.Chair;
import edu.kit.informatik.model.userdata.rooms.Room;
import edu.kit.informatik.repositories.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.sql.Timestamp;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * {@link IModelDtoMapper} für {@link SeatArrangement} and {@link SeatArrangementDto}
 *
 * @author uekai
 * @author ugqbo
 * @version 1.0
 */
@Service
public class SeatArrangementMapper implements IModelDtoMapper<SeatArrangement, SeatArrangementDto> {

    private final UserRepository userRepository;
    private final CourseRepository courseRepository;
    private final RoomRepository roomRepository;
    private final ParticipantRepository participantRepository;
    private final ChairRepository chairRepository;

    /**
     * Konstruktor zum Erstellen eines Objektes der Klasse
     * @param userRepository {@link UserRepository}
     * @param courseRepository {@link CourseRepository}
     * @param roomRepository {@link RoomRepository}
     * @param participantRepository {@link ParticipantRepository}
     * @param chairRepository {@link ChairRepository}
     */
    @Autowired
    public SeatArrangementMapper(UserRepository userRepository,
                                 CourseRepository courseRepository, RoomRepository roomRepository,
                                 ParticipantRepository participantRepository, ChairRepository chairRepository) {
        this.userRepository = userRepository;
        this.courseRepository = courseRepository;
        this.roomRepository = roomRepository;
        this.participantRepository = participantRepository;
        this.chairRepository = chairRepository;
    }

    @Override
    public SeatArrangementDto modelToDto(SeatArrangement seatArrangement) {
        Map<String, String> seatMap = new HashMap<>();
        seatArrangement.getSeatMap().forEach((chair, participant) -> seatMap.put(chair.getId(), participant.getId()));

        return new SeatArrangementDto(
                seatArrangement.getId(),
                seatArrangement.getUser().getId(),
                seatArrangement.getName(),
                seatMap,
                seatArrangement.getRoom().getId(),
                seatArrangement.getCourse().getId(),
                seatArrangement.getCreatedAt(),
                seatArrangement.getUpdatedAt()
        );
    }

    @Override
    public List<SeatArrangementDto> modelToDto(List<SeatArrangement> seatArrangements) {
        List<SeatArrangementDto> seatArrangementDtos = new ArrayList<>();
        seatArrangements.forEach(seatArrangement -> seatArrangementDtos.add(modelToDto(seatArrangement)));

        return seatArrangementDtos;
    }

    @Override
    public SeatArrangement dtoToModel(SeatArrangementDto seatArrangementDto) {
        User user = userRepository.findUserById(seatArrangementDto.getUserId()).orElse(null);
        Course course = courseRepository.findCourseById(seatArrangementDto.getCourseId()).orElse(null);
        Room room = roomRepository.findRoomById(seatArrangementDto.getRoomId()).orElse(null);

        Map<Chair, Participant> seatMap = new HashMap<>();

        seatArrangementDto.getSeatMap().forEach((chairId, participantId) -> {
            Participant participant = participantRepository.findParticipantById(participantId).orElse(null);
            Chair chair = chairRepository.findChairById(chairId).orElse(null);
            seatMap.put(chair, participant);
        });

        Timestamp updatedAt;

        if (seatArrangementDto.getUpdatedAt() == null) {
            updatedAt = seatArrangementDto.getCreatedAt();
        } else {
            updatedAt = seatArrangementDto.getUpdatedAt();
        }

        return new SeatArrangement(user, seatArrangementDto.getName(), room, course,
                                    seatMap, seatArrangementDto.getCreatedAt(), updatedAt);
    }

    @Override
    public List<SeatArrangement> dtoToModel(List<SeatArrangementDto> seatArrangementDtos) {
        List<SeatArrangement> seatArrangements = new ArrayList<>();
        seatArrangementDtos.forEach(seatArrangementDto -> seatArrangements.add(dtoToModel(seatArrangementDto)));

        return seatArrangements;
    }
}

File: Implementierung/frontend/paschi/src/dto/mapper/courses/SeatArrangementMapper.ts
import {IModelDtoMapper} from "@/dto/mapper/IModelDtoMapper";
import {SeatArrangement} from "@/model/userdata/courses/SeatArrangement";
import {SeatArrangementDto} from "@/dto/userdata/courses/SeatArrangementDto";
import {UserController} from "@/controller/UserController";
import {Participant} from "@/model/userdata/interactions/Participant";
import {RoomObject} from "@/model/userdata/rooms/RoomObject";
import {useSeatArrangementStore} from "@/store/SeatArrangementStore";
import {useCourseStore} from "@/store/CourseStore";
import {RoomService} from "@/service/RoomService";
import {CourseService} from "@/service/CourseService";
import {useRoomStore} from "@/store/RoomStore";
import {useStudentStore} from "@/store/StudentStore";
import {useRoomObjectStore} from "@/store/RoomObjectStore";
import {ParticipantService} from "@/service/ParticipantService";

export class SeatArrangementMapper implements IModelDtoMapper<SeatArrangement, SeatArrangementDto> {

  private static mapper: SeatArrangementMapper = new SeatArrangementMapper();

  private constructor() {
  }

  static getMapper(): SeatArrangementMapper {
    return SeatArrangementMapper.mapper;
  }

  modelToDto(arrangement: SeatArrangement): SeatArrangementDto {
    const seatMap: Map<string, string> = new Map<string, string>();
    const roomId = arrangement.room.getId;
    const courseId = arrangement.course.getId;

    arrangement.seatMap.forEach((student: Participant, chair: RoomObject) => {
      seatMap.set(chair.getId, student.getId);
    });

    return new SeatArrangementDto(
      arrangement.getId,
      arrangement.user.getId,
      arrangement.createdAt,
      arrangement.updatedAt,
      arrangement.name,
      Object.fromEntries(seatMap),
      roomId,
      courseId
    );
  }


  async dtoToModel(arrangementDto: SeatArrangementDto): Promise<SeatArrangement> {
    const userController = UserController.getUserController();
    const courseService = CourseService.getService();
    const roomService = RoomService.getService();

    let course = useCourseStore().getCourse(arrangementDto.courseId);
    if (course == undefined) {
      course = await courseService.getById(arrangementDto.courseId);
    }
    let room = useRoomStore().getRoom(arrangementDto.roomId);
    if (room == undefined) {
      room = await roomService.getById(arrangementDto.roomId);
    }

    let arrangement = useSeatArrangementStore().getSeatArrangement(arrangementDto.id);
    if (arrangement == undefined) {
      arrangement = new SeatArrangement(
        arrangementDto.id,
        0,
        userController.getUser(),
        arrangementDto.name,
        course!,
        room!
      );
      arrangement.createdAt = arrangementDto.createdAt;
      arrangement.updatedAt = arrangementDto.updatedAt;
      useSeatArrangementStore().addSeatArrangement(arrangement);
    } else if (arrangement.updatedAt === arrangementDto.updatedAt) {
      return arrangement;
    }

    const map = Object.entries(arrangementDto.seatMap);
    const seatMap: Map<RoomObject, Participant> = new Map<RoomObject, Participant>();
    for (const value of map) {
      let participant = useStudentStore().getParticipant(value[1]);
      if (participant == undefined) {
        participant = await ParticipantService.getService().getById(value[1]);
      }
      const roomObject = useRoomObjectStore().getChair(value[0])!;
      seatMap.set(roomObject, participant!);
    }

    arrangement.seatMap = seatMap;
    return arrangement;
  }
}
